docs/style.md

```diff
@@ -539,6 +539,245 @@ import { unstable_Pagination as Pagination } from 'carbon-components-react';
 This code should be treated as experimental and will break between release
 versions for the package that it is being imported from.
 
+### Testing
+
+#### Strategy
+
+In general we aim to test components from a user-focused perspective. This means
+avoiding testing implementation details, and instead focusing on writing tests
+that closely resemble how the components are used. The various `testing-library`
+packages are used to encourage this mindset when writing and composing test
+suites.
+
+#### Organization
+
+Every component should have tests covering a series of Categories
+
+- General component functionality/API
+- Accessibility
+- End to end tests
+- Server side rendering
+
+Each of these are separated into individual files. In some cases the syntax may
+be slightly different and separate files make this easier to understand.
+Additionally separate file types can be more easily globbed to only run certain
+tests in certain environments (local, CI, Pre-release checks, etc).
+
+| File name                    | Category                        |
+| ---------------------------- | ------------------------------- |
+| `ComponentName-test.js`      | General component functionality |
+| `ComponentName-test.a11y.js` | Accessibility testing           |
+| `ComponentName-test.e2e.js`  | End to end tests                |
+| `ComponentName-test.ssr.js`  | Server side rendering           |
+
+There are corresponding commands to run all categories, individual categories,
+or a combination. Depending on your shell, modifiers can be used to run two
+commands one after another. Refer to the documentation of your shell.
+
+| Command                           | Corresponding test category                                          |
+| --------------------------------- | -------------------------------------------------------------------- |
+| `yarn test`                       | All categories                                                       |
+| `yarn test:unit`                  | Only component unit tests                                            |
+| `yarn test:a11y`                  | Only accessibility tests                                             |
+| `yarn test:e2e`                   | Only end to end tests                                                |
+| `yarn test:ssr`                   | Only server side tests                                               |
+| `yarn test:a11y && yarn test:e2e` | In `bash` via `&&`: Run the a11y tests, and if they succeed, run e2e |
+
+#### Recipes
+
+Below are some common receipes for component testing. Many of the pattern/syntax
+details contained within these recipes are enforced via eslint rules declared in
+`eslint-config-carbon`.
+
+##### `ComponentName-test.js`
```

Comment:
I wonder if we should add a preamble here around what this file is for. I think if we have Cypress then this file decidedly becomes testing the "API" of the component, right?

Reply:
👍  I'll take a stab at it. Based on how slow/expensive cypress is I think it makes sense to encapsulate as much as possible here with only a specific limited set of items in the e2e tests. I'm thinking of the testing pyramid and all - ideally this file would cover the majority of component functionality.

That said, I think it's hard to codify right now without having yet dug into refactoring our existing tests off enzyme.

Reply:
@tay1orjones that sucks about Cypress being so slow 😞  Do you think it'd be worth doing a quick hackathon to see how far we could take a custom playwright setup that works with Jest?

Reply:
@joshblack not opposed! I think that could be a great idea.